File: java/src/main/java/com/cc/tictactoe/TicTacToe.java
package com.cc.tictactoe;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class TicTacToe {
    public static final int BOARD_SIZE = 3;
    private final int[][][] LINES_INDEXES = {
            {{0, 0}, {0, 1}, {0, 2}},
            {{1, 0}, {1, 1}, {1, 2}},
            {{2, 0}, {2, 1}, {2, 2}},
            {{0, 0}, {1, 0}, {2, 0}},
            {{0, 1}, {1, 1}, {2, 1}},
            {{0, 2}, {1, 2}, {2, 2}},
            {{0, 0}, {1, 1}, {2, 2}},
            {{0, 2}, {1, 1}, {2, 0}}
    };
    private final TicTacToeElement[][] board = new TicTacToeElement[3][3];
    private TicTacToeElement startTurn;
    private TicTacToeElement turn;
    private int filled;

    public TicTacToe(TicTacToeElement turn) {
        this.startTurn = turn;
        this.turn = turn;
        this.filled = 0;
    }

    public static TicTacToe clone(TicTacToe ticTacToe) {
        TicTacToe clone = new TicTacToe(ticTacToe.getStartTurn());
        for (int i = 0; i < ticTacToe.board.length; i++) {
            System.arraycopy(ticTacToe.board[i], 0, clone.board[i], 0, ticTacToe.board[i].length);
        }
        clone.turn = ticTacToe.getTurn();
        clone.filled = ticTacToe.filled;
        return clone;
    }

    @Nullable
    public static TicTacToeElement winnerFromGameStatus(GameStatus gameStatus) {
        return switch (gameStatus) {
            case X_WINS -> TicTacToeElement.X;
            case O_WINS -> TicTacToeElement.O;
            default -> null;
        };
    }

    public static TicTacToeElement getOpponent(TicTacToeElement player) {
        return player == TicTacToeElement.X ? TicTacToeElement.O : TicTacToeElement.X;
    }

    public TicTacToeElement[][] getBoard() {
        return board;
    }

    public TicTacToeElement[][] getClonedBoard() {
        TicTacToeElement[][] copy = new TicTacToeElement[3][3];
        for (int i = 0; i < board.length; i++) {
            System.arraycopy(board[i], 0, copy[i], 0, board[i].length);
        }
        return copy;
    }

    public TicTacToeElement getStartTurn() {
        return startTurn;
    }

    public TicTacToeElement getTurn() {
        return turn;
    }

    public int getFilled() {
        return filled;
    }

    public boolean input(int rowIndex, int columnIndex) {
        if (board[rowIndex][columnIndex] == null) {
            board[rowIndex][columnIndex] = turn;
            toggleTurn();
            this.filled++;
            return true;
        }
        return false;
    }

    public void resetBoard(TicTacToeElement turn) {
        startTurn = turn;
        this.turn = turn;
        this.filled = 0;
        for (TicTacToeElement[] row : board) {
            Arrays.fill(row, null);
        }
    }


    @NotNull
    public GameStatus getGameStatus() {
        if (filled < 5) return GameStatus.IN_PROGRESS;
        TicTacToeElement winner = checkWinner();
        if (winner != null) {
            return winner == TicTacToeElement.X ? GameStatus.X_WINS : GameStatus.O_WINS;
        }
        if (filled == 9) return GameStatus.DRAW;
        return GameStatus.IN_PROGRESS;
    }

    public List<int[]> getAvailableMoves() {
        List<int[]> availableMove = new ArrayList<>();
        for (int i = 0; i < TicTacToe.BOARD_SIZE; i++) {
            for (int j = 0; j < TicTacToe.BOARD_SIZE; j++) {
                if (board[i][j] == null) {
                    availableMove.add(new int[]{i, j});
                }
            }
        }
        return availableMove;
    }


    @Nullable
    private TicTacToeElement checkWinner() {
        for (int[][] x : LINES_INDEXES) {
            TicTacToeElement[] cells = {board[x[0][0]][x[0][1]], board[x[1][0]][x[1][1]], board[x[2][0]][x[2][1]]};
            if (cells[0] != null && cells[0] == cells[1] && cells[1] == cells[2]) {
                return cells[0];
            }
        }
        return null;
    }

    private void toggleTurn() {
        turn = turn == TicTacToeElement.X ? TicTacToeElement.O : TicTacToeElement.X;
    }
}

File: ts/src/TicTacToe.ts
import { GameStatus, TicTacToeElement } from "./typing";

export default class TicTacToe {
  public static readonly BOARD_SIZE = 3;

  private readonly LINES_INDEXES: number[][][] = [
    [
      [0, 0],
      [0, 1],
      [0, 2],
    ],
    [
      [1, 0],
      [1, 1],
      [1, 2],
    ],
    [
      [2, 0],
      [2, 1],
      [2, 2],
    ],
    [
      [0, 0],
      [1, 0],
      [2, 0],
    ],
    [
      [0, 1],
      [1, 1],
      [2, 1],
    ],
    [
      [0, 2],
      [1, 2],
      [2, 2],
    ],
    [
      [0, 0],
      [1, 1],
      [2, 2],
    ],
    [
      [0, 2],
      [1, 1],
      [2, 0],
    ],
  ];
  private board: (TicTacToeElement | null)[][] = [
    [null, null, null],
    [null, null, null],
    [null, null, null],
  ];
  private startTurn: TicTacToeElement;
  private turn: TicTacToeElement;
  private filled: number;

  constructor(turn: TicTacToeElement) {
    this.startTurn = turn;
    this.turn = turn;
    this.filled = 0;
  }

  static clone(ticTacToe: TicTacToe): TicTacToe {
    const clone = new TicTacToe(ticTacToe.startTurn);
    clone.board = ticTacToe.getClonedBoard();
    clone.turn = ticTacToe.turn;
    clone.filled = ticTacToe.filled;
    return clone;
  }

  static winnerFromGameStatus(gameStatus: GameStatus): TicTacToeElement | null {
    switch (gameStatus) {
      case GameStatus.X_WINS:
        return TicTacToeElement.X;
      case GameStatus.O_WINS:
        return TicTacToeElement.O;
      default:
        return null;
    }
  }

  static getOpponent(player: TicTacToeElement): TicTacToeElement {
    return player === TicTacToeElement.X ? TicTacToeElement.O : TicTacToeElement.X;
  }

  // Readonly only prevent modification of the board in compile time (not runtime)
  public getBoard(): Readonly<Readonly<(TicTacToeElement | null)[]>[]> {
    return this.board;
  }

  public getClonedBoard(): (TicTacToeElement | null)[][] {
    return this.board.map((row) => [...row]);
  }

  public getStartTurn(): TicTacToeElement {
    return this.startTurn;
  }

  public getTurn(): TicTacToeElement {
    return this.turn;
  }

  public getFilled(): number {
    return this.filled;
  }

  public input(rowIndex: number, columnIndex: number): boolean {
    if (this.board[rowIndex][columnIndex] === null) {
      this.board[rowIndex][columnIndex] = this.turn;
      this.toggleTurn();
      this.filled++;
      return true;
    }
    return false;
  }

  public resetBoard(turn: TicTacToeElement): void {
    this.startTurn = turn;
    this.turn = turn;
    this.filled = 0;
    for (let i = 0; i < TicTacToe.BOARD_SIZE; i++) {
      for (let j = 0; j < TicTacToe.BOARD_SIZE; j++) {
        this.board[i][j] = null;
      }
    }
  }

  public getGameStatus(): GameStatus {
    const winner = this.checkWinner();
    if (winner !== null) {
      return winner === TicTacToeElement.X ? GameStatus.X_WINS : GameStatus.O_WINS;
    }
    if (this.allFilled()) {
      return GameStatus.DRAW;
    }
    return GameStatus.IN_PROGRESS;
  }

  public getAvailableMoves(): number[][] {
    const availableMoves: number[][] = [];
    for (let i = 0; i < TicTacToe.BOARD_SIZE; i++) {
      for (let j = 0; j < TicTacToe.BOARD_SIZE; j++) {
        if (this.board[i][j] === null) {
          availableMoves.push([i, j]);
        }
      }
    }
    return availableMoves;
  }

  private allFilled(): boolean {
    return this.filled === 9;
  }

  private checkWinner(): TicTacToeElement | null {
    for (const x of this.LINES_INDEXES) {
      const cells: (TicTacToeElement | null)[] = [
        this.board[x[0][0]][x[0][1]],
        this.board[x[1][0]][x[1][1]],
        this.board[x[2][0]][x[2][1]],
      ];
      if (cells.every((cell) => cell !== null) && new Set(cells).size === 1) {
        return cells[0];
      }
    }
    return null;
  }

  private toggleTurn(): void {
    this.turn = this.turn === TicTacToeElement.X ? TicTacToeElement.O : TicTacToeElement.X;
  }
}
